pkg/util/metrics_helper_test.go

```diff
@@ -81,3 +82,110 @@ func makeLabels(namesAndValues ...string) []*dto.LabelPair {
 
 	return out
 }
+
+// TestSendSumOfGaugesPerUserWithLabels tests to ensure multiple metrics for the same user with a matching label are
+// summed correctly
+func TestSendSumOfGaugesPerUserWithLabels(t *testing.T) {
```

Comment:
I think this test is too much complicated and quite difficult to follow. I've tried to rewritten it in an easier form. Take a look at the following, please (pay attention that `require.ElementsMatch()` is intentional to avoid flaky tests, because output ordering is not guaranteed to the use of a map).

```
func TestSendSumOfGaugesPerUserWithLabels(t *testing.T) {
	user1Metric := prometheus.NewGaugeVec(prometheus.GaugeOpts{Name: "test_metric"}, []string{"label_one", "label_two"})
	user2Metric := prometheus.NewGaugeVec(prometheus.GaugeOpts{Name: "test_metric"}, []string{"label_one", "label_two"})
	user1Metric.WithLabelValues("a", "b").Set(100)
	user1Metric.WithLabelValues("a", "c").Set(80)
	user2Metric.WithLabelValues("a", "b").Set(60)
	user2Metric.WithLabelValues("a", "c").Set(40)

	user1Reg := prometheus.NewRegistry()
	user2Reg := prometheus.NewRegistry()
	user1Reg.MustRegister(user1Metric)
	user2Reg.MustRegister(user2Metric)

	mf := BuildMetricFamiliesPerUserFromUserRegistries(map[string]*prometheus.Registry{
		"user-1": user1Reg,
		"user-2": user2Reg,
	})

	desc := prometheus.NewDesc("test_metric", "", []string{"user", "label_one"}, nil)
	actual := collectMetrics(func(out chan prometheus.Metric) {
		mf.SendSumOfGaugesPerUserWithLabels(out, desc, "test_metric", "label_one")
	})
	expected := []*dto.Metric{
		{Label: makeLabels("label_one", "a", "user", "user-1"), Gauge: &dto.Gauge{Value: proto.Float64(180)}},
		{Label: makeLabels("label_one", "a", "user", "user-2"), Gauge: &dto.Gauge{Value: proto.Float64(100)}},
	}
	require.ElementsMatch(t, expected, actual)

	desc = prometheus.NewDesc("test_metric", "", []string{"user", "label_two"}, nil)
	actual = collectMetrics(func(out chan prometheus.Metric) {
		mf.SendSumOfGaugesPerUserWithLabels(out, desc, "test_metric", "label_two")
	})
	expected = []*dto.Metric{
		{Label: makeLabels("label_two", "b", "user", "user-1"), Gauge: &dto.Gauge{Value: proto.Float64(100)}},
		{Label: makeLabels("label_two", "c", "user", "user-1"), Gauge: &dto.Gauge{Value: proto.Float64(80)}},
		{Label: makeLabels("label_two", "b", "user", "user-2"), Gauge: &dto.Gauge{Value: proto.Float64(60)}},
		{Label: makeLabels("label_two", "c", "user", "user-2"), Gauge: &dto.Gauge{Value: proto.Float64(40)}},
	}
	require.ElementsMatch(t, expected, actual)

	desc = prometheus.NewDesc("test_metric", "", []string{"user", "label_one", "label_two"}, nil)
	actual = collectMetrics(func(out chan prometheus.Metric) {
		mf.SendSumOfGaugesPerUserWithLabels(out, desc, "test_metric", "label_one", "label_two")
	})
	expected = []*dto.Metric{
		{Label: makeLabels("label_one", "a", "label_two", "b", "user", "user-1"), Gauge: &dto.Gauge{Value: proto.Float64(100)}},
		{Label: makeLabels("label_one", "a", "label_two", "c", "user", "user-1"), Gauge: &dto.Gauge{Value: proto.Float64(80)}},
		{Label: makeLabels("label_one", "a", "label_two", "b", "user", "user-2"), Gauge: &dto.Gauge{Value: proto.Float64(60)}},
		{Label: makeLabels("label_one", "a", "label_two", "c", "user", "user-2"), Gauge: &dto.Gauge{Value: proto.Float64(40)}},
	}
	require.ElementsMatch(t, expected, actual)
}

func collectMetrics(send func(out chan prometheus.Metric)) []*dto.Metric {
	out := make(chan prometheus.Metric)

	go func() {
		send(out)
		close(out)
	}()

	metrics := []*dto.Metric{}
	for m := range out {
		collected := &dto.Metric{}
		m.Write(collected)

		metrics = append(metrics, collected)
	}

	return metrics
}
```

Reply:
👍 Yes, this is much easier to follow. I was confused for a second about the labels order, but it turns out that Prometheus will sort them by name in `prometheus.MustNewConstMetric`, who knew? :) 

Personally I'd put in some extra `{ ... }` blocks, to separate variables between individual tests, but that's really just nitpicking.

Reply:
I will update to use this test.